CRAYSAT-1429: Add documentation for blade swap using SAT (#1654)
Cray-HPE/docs-csm: operations/node_management/Replace_a_Compute_Blade_Using_SAT.md

      | x9000c3s0b3n0 | Node | 3        | Off   | OK   | True    | X86  | River | Compute     | Sling    |
      | x9000c3s0b4n0 | Node | 4        | Off   | OK   | True    | X86  | River | Compute     | Sling    |
      +---------------+------+----------+-------+------+---------+------+-------+-------------+----------+
      ```

   1. Find the `bos_session` value for each node via the Configuration Framework Service (CFS).

      ```bash
      ncn# cray cfs components describe x9000c3s0b1n0 | grep bos_session
      ```

      Example output:

      ```text
      bos_session = "e98cdc5d-3f2d-4fc8-a6e4-1d301d37f52f"
      ```

   1. Find the required `templateUuid` value with BOS.

      ```bash
      ncn# cray bos session describe BOS_SESSION | grep templateUuid
      ```

      Example output:

      ```text
      templateUuid = "compute-nid1-4-sessiontemplate"
      ```

   1. Determine the list of xnames associated with the desired boot session template.

      ```bash
      ncn# cray bos sessiontemplate describe SESSION_TEMPLATE_NAME | grep node_list
      ```

      Example output:

>       ```text

Reviewer: ```suggestion
      ```toml
```

      node_list = [ "x9000c3s0b1n0", "x9000c3s0b2n0", "x9000c3s0b3n0", "x9000c3s0b4n0",]
      ```

1. Shut down the nodes on the target blade.

>    Use the `sat bootsys` command to shut down the nodes on the target blade. Specify the appropriate component name (xname), and BOS templates determined in the prvious step in a comma-separated list.

Reviewer: ```suggestion
   Use the `sat bootsys` command to shut down the nodes on the target blade. Specify the appropriate component name (xname)
   for the slot, and a comma-separated list of the BOS session templates determined in the previous step.
```

   ```bash

>    ncn# BOS_TEMPLATE=cos-2.0.30-slurm-healthy-compute

Reviewer: ```suggestion
   ncn# BOS_TEMPLATES=cos-2.0.30-slurm-healthy-compute
```

>    ncn# sat bootsys shutdown --stage bos-operations --bos-limit x9000c3s0 --recursive --bos-templates BOS_TEMPLATES

Reviewer: ```suggestion
   ncn# sat bootsys shutdown --stage bos-operations --bos-limit x9000c3s0 --recursive --bos-templates $BOS_TEMPLATES
```

   ```

### Use SAT to remove the blade from hardware management

1. Power off the slot and delete blade information from HSM.

>    Use the `sat swap` command to power off the slot and delete the blade's ethernet interfaces and Redfish endpoints from HSM.

Reviewer: ```suggestion
   Use the `sat swap` command to power off the slot and delete the blade's Ethernet interfaces and Redfish endpoints from HSM.
```

   seconds). Failure to do so may damage the equipment.

### Use SAT to add the blade to hardware management

1. Use the `sat swap` command to begin discovery for the blade and add it to hardware management.

   ```bash
   ncn# sat swap blade -a enable x9000c3s0
   ```

### Perform updates and boot the nodes

1. Optional: If necessary, update the firmware. Review the [Firmware Action Service (FAS)](../firmware/FAS_Admin_Procedures.md) documentation.

   ```bash
   ncn# cray fas actions create CUSTOM_DEVICE_PARAMETERS.json
   ```

1. Update the System Layout Service (SLS).

   1. Dump the existing SLS configuration.

      ```bash
      ncn# cray sls networks describe HSN --format=json > existingHSN.json
      ```

   1. Copy `existingHSN.json` to `newHSN.json`, edit `newHSN.json` with the changes, then run the following command:

      ```bash
      ncn# curl -s -k -H "Authorization: Bearer ${TOKEN}" https://API_SYSTEM/apis/sls/v1/networks/HSN \
                -X PUT -d @newHSN.json
      ```

1. Reload DVS on NCNs.

   See *HPE Cray Operating System Administration Guide: CSM on HPE Cray EX Systems (S-8024)* for more information.

1. Power on and boot the nodes.

>    Use `sat bootsys` to power on and boot the nodes. Specify the appropriate BOS template for the node type.

Reviewer: The previous BOS examples show picking out a template based on xnames. In this case you are indicating that they should select the template based on the node type (which makes sense, since this is a new blade, so the node xnames likely won't exist in any existing BOS session templates). However, this means instructions should be provided on how to select a template based on node type.

If I were you, I'd move your current BOS writeup into its own document, and have that document explain how to identify the appropriate BOS session template for a node by xname or by type. That way you don't have to repeat the procedure so many times in these documents.

Alternatively, in the interest of getting this PR merged sooner rather than later, I can open a new ticket for this piece, and you can just leave this "select template for the node type" text as-is. If you want to go that route, let me know, and I'll open a ticket for the BOS stuff. You wouldn't need to work on that ticket, since it's a CSM procedure, not SAT.

Author: "Factoring out" the procedure to get the appropriate BOS session templates is probably the best way forward, as I copied the procedure from the file `operations/power_management/Prepare_the_System_for_Power_Off.md`. 

I should note that there also already exists a procedure for powering off compute nodes and UANs in the file `operations/power_management/Shut_Down_and_Power_Off_Compute_and_User_Access_Nodes.md`, though that procedure is for powering off *all* computes and UANs using the given session template, not necessarily specific to a single blade (i.e. using the `--bos-limit` argument to `sat bootsys`.) That procedure could be modified to accommodate this use case and then this first step could simply be a link that that document, but in the interest of not making too many changes to the structure of the existing docs I decided to keep these changes relatively self-contained.
